crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+                )
+            })?;
+
+        // Check if there is enough time to allocate new full slices
+        let new_full_slices_count = if recent_time > self.min_recent_time {
+            // Rely on integer rounding to get the correct number of full slices
+            // that need adding.
+            (recent_time - self.min_recent_time).as_secs()
+                / ((1u64 << self.factor) * UNIT_TIME.as_secs())
+        } else {
+            0
+        };
+
+        for i in 0..new_full_slices_count {
+            // Store the hash of the full slice
+            let op_hashes = store
+                .retrieve_op_hashes_in_time_slice(
+                    full_slices_end_timestamp,
+                    full_slices_end_timestamp
+                        + Duration::from_secs(
+                            i * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+                        ),
+                )
+                .await?;
+
+            let hash = combine_op_hashes(op_hashes);
+
+            store
+                .store_slice_hash(self.full_slices, hash.to_vec())
+                .await?;
+
+            self.full_slices += 1;
+        }
+
+        full_slices_end_timestamp += Duration::from_secs(
+            new_full_slices_count * (1u64 << self.factor) * UNIT_TIME.as_secs(),
+        );
+
+        let new_partials =
+            self.layout_partials(current_time, full_slices_end_timestamp)?;
+        let old_partials = std::mem::take(&mut self.partial_slices);
+
+        for (start, size) in new_partials.into_iter() {
+            // If this slice didn't change, then we can reuse the hash
+            let maybe_old_hash = old_partials.iter().find_map(|b| {
+                if b.start == start && b.size == size {
+                    Some(b.hash.clone())
+                } else {
+                    None
+                }
+            });
+
+            // Otherwise, we need to get the op hashes for this time slice and combine them
+            let hash = match maybe_old_hash {
+                Some(h) => h,
+                None => {
+                    let end = start
+                        + Duration::from_secs(
+                            (1u64 << size) * UNIT_TIME.as_secs(),
+                        );
+                    combine_op_hashes(
+                        store
+                            .retrieve_op_hashes_in_time_slice(start, end)
+                            .await?,
+                    )
+                }
+            };
+
+            self.partial_slices.push(PartialSlice { start, size, hash })
+        }
+
+        // There will be a small amount of time left over, which we can't partition into smaller
+        // slices. Some amount less than [UNIT_TIME] will be left over.
+        // Once that has elapsed up to the next [UNIT_TIME] boundary, we can update again.
+        if let Some(last_partial) = self.partial_slices.last() {
```

Comment:
Does this mean that the ops in this range don't get gossiped until they're included in a time slice? Or does it mean that we have to send the full list of op hashes in this range to compare with our gossip peer?

Reply:
I am undecided about this. That is what kitsune1 does with its final time slice. So it would certainly be a valid strategy to just say, at the point that we start a gossip round, compute the hash for the time slice between the end of the last partial and the current timestamp.

Alternatively, we don't really need to communicate that last hash. It's a constantly moving target in most cases and we're already just going to eagerly push "what's new since we last spoke". I'd be fine with saying that everything more recent than UNIT_TIME doesn't get hash checked. If we happen to miss something in incremental updates (ideally rare!) then we can find out when those ops make it into a slice.
If we crash and lose some ops that way, then gossip has to try to figure out when to start syncing from anyway, and UNIT_TIME is really a lower bound for that because it would only take one op being created while we were offline to upset that "most recent" hash. So we're going to need to resync at least that window of hashes anyway to get caught up.

What do you think?

Reply:
Either of those approaches seem fine  -- I wasn't sure if the idea was to intentionally *not* gossip ops in the leftover time until they get included in a time slice.